Automattic/jetpack
# Tests: Update Jetpack PHPUnit tests for PHP 8

#### Changes proposed in this Pull Request:


WordPress requires PHPUnit 7, which doesn't support PHP 8. To allow it to run, WordPress core has a local version of the MockObject classes that have been edited to work.

Then we also need to clean up various tests that fail in PHP 8. And I'll throw in a few other cleanups that happened to show up while I was doing that.

Note this PR doesn't fix package tests. They're different and involved enough that I'll put them in a separate PR.

* Fix WP_Test_Jetpack_Sync_Constants::test_does_not_fire_if_constants_havent_changed

  It's calling `constant( $constant )` and trying to catch the PHPUnit exception created from the PHP warning, but in PHP 8 it throws an Error instead of generating a warning.

  The whole idea here seems suspect, if any of the constants are not defined the whole test gets skipped, which makes me wonder how it was ever being run in the first place. More sensible would be to use `defined()` to check each individual constant.

  Also, may as well switch assertEquals() to assertSame() to avoid potential false passes.

* Replace WP_Test_Jetpack_Deprecation::errorHandler

  The $errcontext parameter for error handlers was removed in PHP 8.0.

  Rewrite the test to use the action and filter provided by WordPress instead of trying to use the context.

* Fix WP_Test_Jetpack_Deprecation::test_deprecated_method_smoke_test

  We can't just pass a bogus string anymore, one of the methods starts throwing an error in PHP 8 because array_intersect_key() now throws if not passed arrays.

  On the plus side, that means we can get rid of the "skip to avoid other warnings" thing.

* Fix "risky" WP_Test_Jetpack_Sync_Codec_Interface::test_sync_codec_does_not_explode_on_circular_reference

  Not a PHP 8 issue, PHPUnit has long complained about tests that contain no assertions. But may as well fix it anyway.

* wp_localize_script() only accepts arrays

  It only accidentally kind-of worked when passed strings. Now in PHP 8 it starts issuing a warning when trying to access the string as an array at one point. Fixes #17628.

* Recompress tests/php/modules/photon/sample-content/test-image-medium.png

  The image has an iCCP chunk that causes libpng to issue warnings. This doesn't seem to be intentional as part of any test, so let's just fix it.

  This also takes the image from 22 KB to 3 KB.


#### Jetpack product discussion



#17166

#### Does this pull request change what data or activity we track or use?



No.

#### Testing instructions:






* See if Jetpack tests pass in PHP 8 (note package tests will still fail, that's #17954).
* Check that the Image Compare block's handle still has the aria label.
* Check that "Dynamically set WebPack's publicPath so that split assets can be found" still works. I don't know what that does.
* Check that the site-logo theme tool still sets classes correctly in site-logo-header-text.js.

#### Proposed changelog entry for your changes:


* Perhaps "General: improve compatibility with PHP 8."
